### PR TITLE
Adding advisor Rules and parser scripts with unit tests.

### PR DESCRIPTION
**Pull request by poojam23**
Summary: This adds some rules in the tools/advisor/advisor/rules.ini file and corresponding python parser scripts for parsing the rules file and the rocksdb LOG and OPTIONS files. WIP for adding rules depending on ODS. The starting point of the script is the rocksdb/tools/advisor/advisor/rule_parser.py file.

Test Plan: Unit tests for the parser have been added to the rocksdb/tools/advisor/test directory.

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: tools/advisor/advisor/db_options_parser.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+                    self.options_dict[option_prefix + '.' + key] = value
+                else:
+                    error = 'Not able to parse line in Options file.'
+                    OptionsSpecParser.exit_with_parse_error(line, error)
+
+    def check_and_trigger_conditions(self, conditions):
+        # for every condition, if the fields are not present set_trigger will
+        # not be called for it. Or if all the fields are present, then the
+        # trigger will be set to whatever the expression evaluates to.
+        self.load_from_source()
```
Comment (maysamyabandeh) [inline tools/advisor/advisor/db_options_parser.py]: do we need to load it each time we call check_*? Can it be loaded once and used for every other calls?

File: tools/advisor/advisor/db_options_parser.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        for cond in conditions:
+            prefix_ix = []
```
Comment (maysamyabandeh) [inline tools/advisor/advisor/db_options_parser.py]: Give a proper name (and perhaps with a comment) to explain that the this list of options that will need to be concatenated to the prefix provided in the rule.

File: tools/advisor/advisor/db_options_parser.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                    options.append(0)
+                ix += 1
+
+            # if all the options were present as is:
+            if not prefix_ix:
+                if not eval(cond.eval_expr):
+                    cond.set_trigger(cond.eval_expr)
+                continue
+
+            # for all the options that were not present as is:
```
Comment (maysamyabandeh) [inline tools/advisor/advisor/db_options_parser.py]: add more comment: "then we apply the option to every column family found in option file"

File: tools/advisor/advisor/ini_parser.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+from enum import Enum
```
Comment (maysamyabandeh) [inline tools/advisor/advisor/ini_parser.py]: Add facebook headers to all files. e.g., https://github.com/facebook/rocksdb/blob/master/utilities/transactions/write_prepared_txn.h

File: tools/advisor/advisor/ini_parser.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        line = line.strip()
+        comment_start = line.find('#')
+        if comment_start > -1:
+            return line[:comment_start]
+        return line
+
+    @staticmethod
+    def is_section_header(line):
+        line = line.strip()
+        if line.startswith('[') and line.endswith(']'):
```
Comment (maysamyabandeh) [inline tools/advisor/advisor/ini_parser.py]: add a sample line as an inline comment to clarify what the code does.

File: tools/advisor/advisor/ini_parser.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            return True
+        return False
+
+    @staticmethod
+    def get_section_name(line):
+        token_list = line.strip()[1:-1].split('"')
```
Comment (maysamyabandeh) [inline tools/advisor/advisor/ini_parser.py]: same here.

File: tools/advisor/advisor/rule_parser.py [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+        return self.data_source
+
+    def reset_trigger(self):
+        self.trigger = None
+
+    def set_trigger(self, condition_trigger):
+        self.trigger = condition_trigger
+
+    def is_triggered(self):
+        if self.trigger:
```
[the remark this answers is not in the document]
Comment (maysamyabandeh) [reply tools/advisor/advisor/rule_parser.py]: what do you think about this comment?

[the remark this answers is not in the document]
Comment (poojam23) [reply tools/advisor/advisor/rule_parser.py]: self.trigger is either a log or a map or a boolean depending on the kind of condition and it is set to a value only if the condition is satisfied. I wanted to keep the is_triggered() method separate from the get_trigger(), hence I return a boolean here specifically.

**Pull request closed without merging**